go-controller/pkg/ovn/obj_retry.go

```diff
@@ -572,7 +597,18 @@ func (oc *Controller) addResource(objectsToRetry *retryObjs, obj interface{}, fr
 			extraParameters.portGroupIngressDenyName,
 			extraParameters.portGroupEgressDenyName,
 			obj)
-
+	case factory.EgressFirewallType:
+		egressFirewall := obj.(*egressfirewall.EgressFirewall).DeepCopy()
+		if err := oc.addEgressFirewall(egressFirewall); err != nil {
+			egressFirewall.Status.Status = egressFirewallAddError
+		} else {
+			egressFirewall.Status.Status = egressFirewallAppliedCorrectly
+		}
+		if err := oc.updateEgressFirewallStatusWithRetry(egressFirewall); err != nil {
+			klog.Error(err)
+		}
+		metrics.UpdateEgressFirewallRuleCount(float64(len(egressFirewall.Spec.Egress)))
+		metrics.IncrementEgressFirewallCount()
```

Comment:
Hmm... so now we're not returning the error any more. If, for whatever reason, the add operation fails, the egressfirewall status will be updated and that's it. The add won't be retried. 

If we do return an error, here we actually have two very distinct errors to take care of:

- the one from the actual resource add, which we should care about, 
- the one from the status update, which comes from the API server and... which we should discuss how to handle!

We have three possible error situations: 

1. add failed, status update succeeded 
2. add succeeded, status updated failed
3. add failed, status update failed. 

If we do want to include the status update here (vs a separate function as Mohamed previously did), we'll have to consider this as one retriable block and aggregate the two errors, so that the three cases above are considered the same and they all trigger a retry.

@trozet what's your take on this? :)

Reply:
yeah I should aggr both errors and return them

Reply:
so one thing I notice (with this patch and the old code) is that the metrics are alwasy Incremented/Updated even if the add fails. Same behavior fore decrementing/delete. Should we not touch the metrics if there is an error adding the firewall? Seems like we should still add them if egressfirewall add works, but status update fails. @martinkennelly wdyt?

Reply:
Looking at the code that actually adds an egressfirewall, we get an error if the firewall exists already:

```
func (oc *Controller) addEgressFirewall(egressFirewall *egressfirewallapi.EgressFirewall) error {
	klog.Infof("Adding egressFirewall %s in namespace %s", egressFirewall.Name, egressFirewall.Namespace)

	ef := cloneEgressFirewall(egressFirewall)
	ef.Lock()
	defer ef.Unlock()
	// there should not be an item already in egressFirewall map for the given Namespace
	if _, loaded := oc.egressFirewalls.LoadOrStore(egressFirewall.Namespace, ef); loaded {
		return fmt.Errorf("error attempting to add egressFirewall %s to namespace %s when it already has an egressFirewall",
			egressFirewall.Name, egressFirewall.Namespace)
	}
```
So in error situation 2 from my previous message (add succeeded, status update failed), the following will happen:
- egress firewall is correctly added
- status update returns an error, so `addResource` returns this error, causing the retry entry for this egress firewall to be marked for addition later on
- then, `iterateRetryResources` will pick this entry and will try to add the egress firewall again
- `addEgressFirewall` will fail since the egress firewall exists already
- the status update can succeeded or fail, but `addResource` will still return the error we got from `addEgressFirewall`
- `iterateRetryResources` will try to add this egress firewall over and over, regardless of the outcome the status update, since  `addEgressFirewall`  will always return an error.

I'm starting to think that we should probably decouple the add from the status update... At the very least, we can keep both operations here and only return the error from `addEgressFirewall`. If, as Tim was hinting at, we expect the status update to fail very rarely, we might even claim that the current implementation is good enough already since it does some retrying:
```
func (oc *Controller) updateEgressFirewallWithRetry(egressfirewall *egressfirewallapi.EgressFirewall) error {
	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
		return oc.kube.UpdateEgressFirewall(egressfirewall)
	})
	if retryErr != nil {
		return fmt.Errorf("error in updating status on EgressFirewall %s/%s: %v",
			egressfirewall.Namespace, egressfirewall.Name, retryErr)
	}
	return nil
}
```
In case the status update keeps failing, I would think there's a serious problem with the API server. Would it be acceptable to say that the cluster admin will have to delete and re-add the egress firewall once the API server is back?